dwmblocks.1
.TH dwmblocks2 1 "2023"
.SH NAME
dwmblocks2 \- Lightweight status bar for dwm, a fork of dwmblocks (not a drop-in replacement)

.SH SYNOPSIS
.B dwmblocks2

.SH DESCRIPTION
.I dwmblocks2
is a lightweight status bar for the
.I dwm
window manager. It is a fork of the original dwmblocks, designed to provide an efficient and customizable status bar. 
.I dwmblocks2
allows you to display various information in your dwm status bar, such as system statistics, the time, and more. This man page provides an overview of how to use and configure 
.I dwmblocks2.

.SH USAGE
To run
.I dwmblocks2,
add the following line to your
.I xinitrc
or startup script:
.PP
.B dwmblocks2 &
.PP
This command will start
.I dwmblocks2
in the background.

.SH CONFIGURATION
Configuration for
.I dwmblocks2
is done by editing the
.I blocks.h
file. By default,
.I dwmblocks2
uses all blocks defined in the
.I blocks/
directory, in addition to the built-in clock block. You can customize the blocks to display external programs or C functions.

.TP
.B Block Configuration
Blocks are defined in the
.I blocks.h
file. Each block is represented as a command, but it is not interpreted by the shell. Instead, it is likely a script or program. The overhead and loss of readability due to shell interpretation are avoided for efficiency.

.TP
.B Signal Numbers
A signal number, specified through an environmental variable, is mandatory for each block. These signals allow
.I dwmblocks2
to update specific blocks. Signal numbers must be unique and fall within the range of 1 to SIGRTMAX - SIGRTMIN. You can set the same signal for multiple blocks if needed.

.TP
.B Block Updates
Blocks whose interval is set to 0 will only be updated through signals, including those sent by
.I dwm
through clicks. This allows for efficient and independent updates of blocks.

.SH CLICKS
For mouse clicks to work with
.I dwmblocks2,
you need to apply the
.I dwm-statuscmd.diff
patch to your
.I dwm
installation. The button number clicked is passed as the first argument to the block, and
.I dwmblocks2
will update the block accordingly.

.PP
The following click events are supported:
.TP
.B 0
No click (other signal/interval)
.TP
.B 1
Left button click
.TP
.B 2
Middle button click
.TP
.B 3
Right button click
.TP
.B 4
Scroll up
.TP
.B 5
Scroll down
.TP
.B 6
Shift + left button click
.TP
.B 7
Control + left button click
.PP
You can customize the action taken in response to these click events by modifying your block scripts.

.SH INSTALLATION
To install
.I dwmblocks2,
follow these steps:
.PP
1. Clone the
.I dwmblocks2
repository from GitHub:
.PP
.B $ git clone https://github.com/lucas-mior/dwmblocks2
.PP
2. Navigate to the
.I dwmblocks2
directory:
.PP
.B $ cd dwmblocks2
.PP
3. Build and install
.I dwmblocks2:
.PP
.B $ make
.B $ sudo make install
.PP
After installation, make sure to add the block scripts to your
.I PATH
and set the environmental variables according to your block configuration in the
.I blocks.h
file. Note that some default blocks may call programs that need to be installed on your system.

.SH DIFFERENCES FROM ORIGINAL DWMBLOCKS
.I dwmblocks2
offers several improvements and differences compared to the original dwmblocks:

.TP
.B \- It's much lighter, since the shell is avoided on every command spawned.
.TP
.B \- A hard-to-reproduce bug in which
.I dwmblocks
would freeze has been fixed.
.TP
.B \- Blocks are updated independently (a slow block will not block others).
.TP
.B \- Clicks are passed through the first argument, not the BLOCK_BUTTON environment variable, which makes it easier to debug scripts.
.TP
.B \- Signals are mandatory and set through environmental variables, so one can more easily keep dwmblocks signals in sync with scripts which update the contents of the bar.
.TP
.B \- There is the possibility of using C functions as blocks, which is useful for the included clock block, which runs every second, avoiding the overhead of forking and executing a new process.

.SH Why poll and not epoll?
The file descriptors watched in
.I dwmblocks2
change all the time, so if
.I epoll
was used, there would be lots of system calls to update them. Using
.I poll,
it's only a matter of assigning to the poll events array.
.I epoll
is meant to be used for handling a huge number of ideally fixed connections, not a simple program reading from some pipes. Besides,
.I poll
is more portable.

.SH LICENSE
.I dwmblocks2
is licensed under GPLv2, and the block scripts in the
.I blocks/
directory are licensed under AGPL.

.SH SEE ALSO
.BR dwm (1)


dwmblocks.c
#include "dwmblocks.h"
#include "blocks.h"

Display *display;
Window root;
static char status_bar[LENGTH(blocks)][BLOCK_OUTPUT_LENGTH] = {0};

char clock_output[BLOCK_OUTPUT_LENGTH] = {0};
int clock_signal;

static char status_new[sizeof (status_bar) + sizeof (clock_output)];
static char status_old[sizeof (status_bar) + sizeof (clock_output)];
static const char delim = ' ';

void get_block_output(const Block *block, char *output) {
    FILE *command_pipe;
    char *status;
    int error;
    usize length;

    if (block->signal) {
        // this messages dwm which signal belongs to which block
        output[0] = (char) block->signal;
        output += 1;
    }
    if (!(command_pipe = popen_no_shell(block->command))) {
        fprintf(stderr, "Failed to run %s: %s\n",
                         block->command, strerror(errno));
        output[0] = '\0';
        return;
    }
    do {
        errno = 0;
        status = fgets(output, BLOCK_OUTPUT_LENGTH - 1, command_pipe);
        error = errno;
    } while (!status && error == EINTR);
    // TODO: Check if pclose() is right here, because
    // popen_no_shell uses pipe() and fdopen()
    pclose(command_pipe);

    length = strcspn(output, "\n");
    output[length] = '\0';
    if (length == 0)
        return;

    while (output[length - 1] == delim) {
        output[length - 1] = '\0';
        length -= 1;
        if (length == 1)
            break;
    }
    if (length > 0) {
        output[length] = delim;
        output[length + 1] = '\0';
    }
    return;
}

void get_block_outputs(int64 seconds) {
    for (uint i = 0; i < LENGTH(blocks); i += 1) {
        Block *block = &blocks[i];
		if (seconds < 0) {
            get_block_output(block, status_bar[i]);
			continue;
		}
		if (block->interval == 0)
			continue;
        if ((seconds % block->interval) == 0)
            get_block_output(block, status_bar[i]);
    }
    block_clock(0);
    return;
}

void set_root(bool check_changed) {
    if (check_changed)
        memcpy(status_old, status_new, sizeof (status_new));

    status_new[0] = '\0';
    for (uint i = 0; i < LENGTH(blocks); i += 1)
        strcat(status_new, status_bar[i]);
    strcat(status_new, clock_output);

    if (check_changed && !memcmp(status_old, status_new, sizeof (status_new))) {
        fprintf(stderr, "Status bar not changed!\n");
        return;
    }

    XStoreName(display, root, status_new);
    XFlush(display);
    return;
}

void signal_handler(int signum) {
    for (uint i = 0; i < LENGTH(blocks); i += 1) {
        Block *block = &blocks[i];
        if (block->signal == (signum - SIGRTMIN))
            get_block_output(block, status_bar[i]);
    }
    set_root(true);
    return;
}

void button_handler(int signum, siginfo_t *signal_info, void *ucontext) {
    char button[2] = {'0' + (signal_info->si_value.sival_int & 7), '\0'};
    Block *block = NULL;
    char *command[2];
    pid_t child;
    (void) ucontext;

    signum = (signal_info->si_value.sival_int >> 3) - SIGRTMIN;
    for (uint i = 0; i < LENGTH(blocks); i += 1) {
        if (blocks[i].signal == signum) {
            block = &blocks[i];
            break;
        }
    }
    if (!block) {
        if (signum == clock_signal)
            block_clock(atoi(button));
        else
            fprintf(stderr, "No block configured for signal %d\n", signum);
        return;
    }

    switch ((child = fork())) {
    case 0:
        command[0] = block->command;
        command[1] = NULL;
        setenv("BLOCK_BUTTON", button, 1);
        execvp(command[0], command);
        fprintf(stderr, "running %s failed: %s\n", command[0], strerror(errno));
        exit(EXIT_SUCCESS);
    case -1:
        fprintf(stderr, "fork failed: %s\n", strerror(errno));
        return;
    default:
        // wait is supposed to fail because of signal_child_action
        waitpid(child, NULL, 0);
        kill(getpid(), SIGRTMIN + block->signal);
    }
	return;
}

FILE *popen_no_shell(char *command) {
    int pipefd[2];
    pid_t pid;

    char *c = command;
    char *argv[4] = { command, NULL };

    if (pipe(pipefd) < 0) {
        perror("pipe");
        return NULL;
    }

    while (*c) {
        if ((*c == ' ') || (*c == '\t')) {
			argv[0] = "/bin/sh";
			argv[1] = "-c";
			argv[2] = command;
			argv[3] = NULL;
            break;
        }
        c += 1;
    }

    switch ((pid = fork())) {
    case 0:
        close(pipefd[0]);
        dup2(pipefd[1], STDOUT_FILENO);
        close(pipefd[1]);

        execvp(argv[0], argv);
        fprintf(stderr, "Error executing %s: %s\n",
                        argv[0], strerror(errno));
        exit(EXIT_FAILURE);
    case -1:
        fprintf(stderr, "Error forking: %s\n", strerror(errno));
        return NULL;
    default:
        close(pipefd[1]);
        return fdopen(pipefd[0], "r");
    }
}

void block_clock(int button) {
    time_t seconds_since_epoch;
    struct tm t;
    char *week;
    char *output = clock_output + 1;

    seconds_since_epoch = time(NULL);
    t = *localtime(&seconds_since_epoch);
    week = ((char *[]) {
        "dom", 
        "seg",
        "ter",
        "qua",
        "qui",
        "sex",
        "sáb"
    })[t.tm_wday];

    snprintf(output, BLOCK_OUTPUT_LENGTH - 1, "📅 %s %02d/%02d %02d:%02d:%02d\n",
             week, t.tm_mday, t.tm_mon + 1, t.tm_hour, t.tm_min, t.tm_sec);

    switch (button) {
    case 1:
        if (fork() == 0) {
            system("yad --calendar --undecorated --fixed --no-buttons "
                   "| tr -d '\n' | xsel -b");
            exit(EXIT_SUCCESS);
        }
        break;
    case 2:
        if (fork() == 0) {
            execlp("print_screen.sh", "print_screen.sh", "tela", NULL);
            exit(EXIT_SUCCESS);
        }
        break;
    case 3:
        if (fork() == 0) {
            execlp("killall", "killall", "yad", NULL);
            exit(EXIT_SUCCESS);
        }
        break;
    default:
        break;
    }
    return;
}
